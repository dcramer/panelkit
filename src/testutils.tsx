import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
