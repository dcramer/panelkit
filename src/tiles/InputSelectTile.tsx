import Tile, { TileProps } from "./Tile";

type InputSelectTileProps = TileProps & {
  entityId: string;
};

export default class InputSelectTile extends Tile<InputSelectTileProps> {
  onTouch = () => {
    const {
      state,
      attributes: { options },
    } = this.getEntity(this.props.entityId);
    const curOption = options.indexOf(state) || -1;
    const nextOption =
      curOption >= options.length - 1 ? options[0] : options[curOption + 1];
    this.callService("input_select", "select_option", {
      entity_id: this.props.entityId,
      option: nextOption,
    });
  };
}
