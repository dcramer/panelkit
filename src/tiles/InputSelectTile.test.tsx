import { renderTile } from "../testutils";
import InputSelectTile from "./InputSelectTile";

test("renders", async () => {
  const { container, getByText } = renderTile(InputSelectTile, {
    entity_id: "input_select.test",
    state: "Afternoon",
    attributes: {
      friendly_name: "Test Input Select",
      editable: false,
      icon: "mdi:house",
      options: ["Morning", "Afternoon", "Evening", "Sleep"],
    },
  });

  expect(getByText(/Test Input Select/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
