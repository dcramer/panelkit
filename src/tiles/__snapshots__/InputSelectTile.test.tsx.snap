// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile state-Afternoon touchable"
  data-testid="tile"
  style="cursor: pointer;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-status"
    >
      Afternoon
    </div>
    <div
      class="tile-body"
    >
      <span
        class="tile-icon"
      >
        <svg />
      </span>
    </div>
    <div
      class="tile-title"
    >
      Test Input Select
    </div>
  </div>
</div>
`;
